buildSrc/src/main/java/org/elasticsearch/gradle/NoticeTask.java

```diff
@@ -0,0 +1,173 @@
+        return inputFile;
+    }
+
+    public void setInputFile(File inputFile) {
+        this.inputFile = inputFile;
+    }
+
+    @OutputFile
+    public File getOutputFile() {
+        return outputFile;
+    }
+
+    public void setOutputFile(File outputFile) {
+        this.outputFile = outputFile;
+    }
+
+    public void licensesDir(File licensesDir) {
+        licensesDirs.add(licensesDir);
+    }
+
+    public List<File> getLicensesDirs() {
+        return Collections.unmodifiableList(this.licensesDirs.stream()
+                .map(file -> new File(file.toString())).collect(Collectors.toList()));
+    }
+    /**
+     * Add notices from the specified directory.
+     */
+
+    public NoticeTask() {
+        setDescription("Create a notice file from dependencies");
+        // Default licenses directory is ${projectDir}/licenses (if it exists)
+        File licensesDir = new File(getProject().getProjectDir(), "licenses");
+
+        if (licensesDir.exists()) {
+            licensesDirs.add(licensesDir);
+        }
+    }
+
+    @TaskAction
+    public void generateNotice() throws IOException {
+        final StringBuilder output = new StringBuilder();
+
+
+        try (Stream<String> lines = Files.lines(this.inputFile.toPath(), StandardCharsets.UTF_8)) {
+            lines.forEach(s -> {
+                output.append(s.trim());
+                output.append("\n");
+            });
+        }
+        output.append("\n\n");
+
+        // This is a map rather than a set so that the sort order is the 3rd
+        // party component names, unaffected by the full path to the various files
+        final Map<String, File> seen = new TreeMap<>();
+
+        licensesDirs.forEach(file -> {
+            try (Stream<Path> pathStream = Files.walk(file.toPath())) {
+                pathStream
+//                    .filter(path -> path.endsWith("-NOTICE.txt"))
+                    .filter(path -> path.toString().endsWith("-NOTICE.txt"))
+                    .map(Path::toFile)
+                    .forEach(licenseFile -> {
+                        // Here we remove the "-NOTICE.txt" so the name can be used later for the -LICENSE.txt file
+                        final String name =
+                            licenseFile.getName().substring(0, licenseFile.getName().length() - "-NOTICE.txt".length());
+
+                        if (seen.containsKey(name)) {
+                            File prevLicenseFile = seen.get(name);
+                            if (getText(prevLicenseFile).equals(getText(licenseFile)) == false) {
+                                throw new RuntimeException("Two different notices exist for dependency '" +
+                                    name + "': " + prevLicenseFile + " and " + licenseFile);
+                            }
+                        } else {
+                            seen.put(name, licenseFile);
+                        }
+                    });
+            } catch (IOException e) {
```

Comment:
No need to catch this here.

Reply:
These are caught as checked and re-thrown as unchecked because they are being handled inside a lambda expression inside a forEach() on the list